packages/dds/sequence/src/intervalCollection.ts

```diff
@@ -1156,6 +1161,116 @@ class EndpointIndex<TInterval extends ISerializableInterval> implements Interval
 	}
 }
 
+export class EndInRangeIndex<TInterval extends ISerializableInterval>
+	implements IntervalIndex<TInterval>
+{
+	private readonly intervalTree;
+
+	constructor(
+		private readonly helpers: IIntervalHelpers<TInterval>,
+		private readonly client?: Client,
+	) {
+		// eslint-disable-next-line @typescript-eslint/unbound-method
+		this.intervalTree = new RedBlackTree<TInterval, TInterval>(helpers.compareEnds);
+	}
+
+	public add(interval: TInterval): void {
+		this.intervalTree.put(interval, interval);
+	}
+
+	public remove(interval: TInterval): void {
+		this.intervalTree.remove(interval);
+	}
+
+	/**
+	 * @returns an array of all intervals contained in this collection whose endpoints locate in the range [start, end)
+	 */
+	public findIntervalsWithEndInRange(start: number, end: number) {
+		if (start <= 0 || start > end - 1 || this.intervalTree.isEmpty()) {
+			return [];
+		}
+		const results: TInterval[] = [];
+		const action: PropertyAction<TInterval, TInterval> = (node) => {
+			results.push(node.key);
+			return true;
+		};
+		const transientStartInterval = this.helpers.create(
+			"transient",
+			start,
+			start,
+			this.client ?? (undefined as any as Client),
+			IntervalType.Transient,
+		);
+
+		const transientEndInterval = this.helpers.create(
+			"transient",
+			end - 1,
```

Comment:
@Abe27342 After testing some corner cases, it has been determined that the `mapRange` function of the `rbtree` will consider the end value as inclusive. To account for this behavior, I am currently subtracting 1 from the end value, not sure if there exists a better approach available

Reply:
subtracting 1 isn't reasonable, since this index should work for regular intervals too (which don't need to be integers). I think the simplest solution is to just make the API technically be inclusive on both ends. Now that I think about it, this probably conceptually behaves as if it's exclusive at the end due to the same reason explained in the remarks doc comment of `SequenceInterval`